In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [4]:
path = r'C:\Users\matth\05-2023 Instacart Basket Analysis'

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [7]:
df_ordsprods_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'order_products__prior.csv'), index_col = False)

In [8]:
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [5]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week   hour_of_day  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

In [6]:
# Create a dataframe

df_test = pd.DataFrame()

In [7]:
# C reate a mixed type column

df_test['mix'] = ['a','b',1,True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# check for mix type data
# weird variable checks if the data type for columns are consistent - True = inconsistent / False = consistent
# print  col prints the problematic column


for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
## how to update data type for a column

df_test['mix'] = df_test['mix'].astype('str')

In [11]:
## assigning is null to the dataframe

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
## to view missing values in details based on 16 above blanks

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [ ]:
## assigning the mean for missing values

df['column with missings'].fillna(mean value, inplace=True)

In [ ]:
## assigning the median for missing values

df['column with missings'].fillna(median value, inplace=True)

In [14]:
## comparing the number of rows in my dataframe

df_prods.shape

(49693, 5)

In [15]:
## creating a dataframe for clean data set, changing the isnull to false

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

In [ ]:
## another code example to drop missing values from field
## inplace = True saves you having to create a new dataframe but override existing one

df_prods.dropna(subset = [‘product_name’], inplace = True)

In [17]:
## looks for duplicates in dataframe

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
## creating a new dataframe not including the duplicate

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

###### OrdsProds Prior Consistency Checks

In [10]:
## assigning is null to the dataframe

df_ordsprods_prior.shape

(32434489, 4)

In [11]:
df_ordsprods_prior.isnull().sum()

order_id             0
product_id           0
add_to_cart_order    0
reordered            0
dtype: int64

In [17]:
## checking for any dups

df_ordsprodsprior_dups = df_ordsprods_prior[df_ordsprods_prior.duplicated()]

In [18]:
df_ordsprodsprior_dups

Empty DataFrame
Columns: [order_id, product_id, add_to_cart_order, reordered]
Index: []

###### Customer consistency checks

In [12]:
## checking the rows and columns of data set

df_cust.shape

(206209, 10)

In [13]:
## checking the number of null values in the customer data set
df_cust.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [15]:
## creating a clean data set removing null first names

df_cust_clean = df_cust[df_cust['First Name'].isnull() == False]

In [16]:
df_cust_clean.shape

(194950, 10)

In [19]:
## checking for any dups

df_cust_clean_dups = df_cust_clean[df_cust_clean.duplicated()]

In [20]:
df_cust_clean_dups

Empty DataFrame
Columns: [user_id, First Name, Surnam, Gender, STATE, Age, date_joined, n_dependants, fam_status, income]
Index: []

### Exercise 4.5.2 -  Nothing about the above dataset looks unusual and requires further investigation

In [22]:

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week   hour_of_day  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

### Exercise 4.5.3 - There are no mixed-type data identified in the below query

In [24]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### Exercise 4.5.5 - there are 206,209 missing values in the days_since_prior_order column. This is most likely due to the fact that orders have actually not been placed yet for the relevant products

In [25]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_of_day                    0
days_since_prior_order    206209
dtype: int64

### Exercise 4.5.6 - decide to remove those blank and only focus on order that have actually been placed with Intacart, keeping the data focused & targetted

In [26]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [27]:
df_ords.shape

(3421083, 7)

In [28]:
df_ords_clean.shape

(3214874, 7)

### Exercise 4.5.7 - check for duplicates complete

In [29]:
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [30]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, hour_of_day, days_since_prior_order]
Index: []

### Exercise 4.5.8 - no duplicates identified as per above code output

In [ ]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))